Introduce isTensorSubclassLike; add special cases to backwards formulas

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#69534 Introduce isTensorSubclassLike; add special cases to backwards formulas**

Something is TensorSubclassLike if it is a Tensor subclass or if it has
the same problems as Tensor subclasses. Today that just includes Tensor
Subclasses and meta tensors but may include other things in the future.

Some of our backwards formulas are incompatible with TensorSubclassLike
objects. For example, calling .data_ptr() is a problem because many
TensorSubclassLike objects don't have storage. Another problem is
in-place operations: performing `regular_tensor.inplace_(tensor_subclass)`
is a problem.

This PR adds special cases to the backward formulas for torch.max and
torch.clamp to handle this. The backward formulas for torch.max and
torch.clamp are not dispatcher operations so they cannot be overridden
and we hesitate to make them dispatcher operations for FC/BC concerns
and performance overhead concerns.

Furthermore, the old concept of "is this inplace operation vmap
compatible?" can be subsumed by the general "is this inplace operation
tensor-subclass compatible" question, so I replaced all instances of
isInplaceVmapCompatible and replaced it with the isTensorSubclassLike
checks.

Test Plan
- I tested the changes using functorch.
- It's possible to write a test for these in core (one has to make
a custom tensor subclass and then send it through the operation and then
invoke autograd), but I wanted to push the work to doing some
generic testing for backward formulas
(https://github.com/pytorch/pytorch/issues/69530) instead of doing some
one-off things now.

Differential Revision: [D32967727](https://our.internmc.facebook.com/intern/diff/D32967727)

zou3519 committed:
Introduce isTensorSubclassLike; add special cases to backwards formulas

Something is TensorSubclassLike if it is a Tensor subclass or if it has
the same problems as Tensor subclasses. Today that just includes Tensor
Subclasses and meta tensors but may include other things in the future.

Some of our backwards formulas are incompatible with TensorSubclassLike
objects. For example, calling .data_ptr() is a problem because many
TensorSubclassLike objects don't have storage. Another problem is
in-place operations: performing `regular_tensor.inplace_(tensor_subclass)`
is a problem.

This PR adds special cases to the backward formulas for torch.max and
torch.clamp to handle this. The backward formulas for torch.max and
torch.clamp are not dispatcher operations so they cannot be overridden
and we hesitate to make them dispatcher operations for FC/BC concerns
and performance overhead concerns.

Furthermore, the old concept of "is this inplace operation vmap
compatible?" can be subsumed by the general "is this inplace operation
tensor-subclass compatible" question, so I replaced all instances of
isInplaceVmapCompatible and replaced it with the isTensorSubclassLike
checks.

Test Plan
- I tested the changes using functorch.
- It's possible to write a test for these in core (one has to make
a custom tensor subclass and then send it through the operation and then
invoke autograd), but I wanted to push the work to doing some
generic testing for backward formulas
(https://github.com/pytorch/pytorch/issues/69530) instead of doing some
one-off things now.

[ghstack-poisoned]

zou3519 committed:
Update on "Introduce isTensorSubclassLike; add special cases to backwards formulas"


Something is TensorSubclassLike if it is a Tensor subclass or if it has
the same problems as Tensor subclasses. Today that just includes Tensor
Subclasses and meta tensors but may include other things in the future.

Some of our backwards formulas are incompatible with TensorSubclassLike
objects. For example, calling .data_ptr() is a problem because many
TensorSubclassLike objects don't have storage. Another problem is
in-place operations: performing `regular_tensor.inplace_(tensor_subclass)`
is a problem.

This PR adds special cases to the backward formulas for torch.max and
torch.clamp to handle this. The backward formulas for torch.max and
torch.clamp are not dispatcher operations so they cannot be overridden
and we hesitate to make them dispatcher operations for FC/BC concerns
and performance overhead concerns.

Furthermore, the old concept of "is this inplace operation vmap
compatible?" can be subsumed by the general "is this inplace operation
tensor-subclass compatible" question, so I replaced all instances of
isInplaceVmapCompatible and replaced it with the isTensorSubclassLike
checks.

Test Plan
- I tested the changes using functorch.
- It's possible to write a test for these in core (one has to make
a custom tensor subclass and then send it through the operation and then
invoke autograd), but I wanted to push the work to doing some
generic testing for backward formulas
(https://github.com/pytorch/pytorch/issues/69530) instead of doing some
one-off things now.

[ghstack-poisoned]

albanD commented on aten/src/ATen/TensorSubclassLikeUtils.h:
```diff
@@ -0,0 +1,48 @@
+// - a Tensor subclass (via __torch_dispatch__ in Python or extending
+//   TensorImpl in C++)
+// - anything else that shares the same perils as Tensor subclasses.
+//   For example, many Tensor subclasses do not have storage and meta Tensors
```
This should also include sparse no? And we can use this to replace sparse checks in a bunch of places (thinking about the autograd engine)

zou3519 replied:
Should it include sparse? Sparse tensors have additional constraints: in addition to (1) no data_ptr and (2) no incompatible in-place, sparse tensors also have restrictions around what view operations may be applied to them.

albanD replied:
So they have even more restrictions? they should at least be here right?

zou3519 replied:
Yes, they have even more restrictions :(. We can certainly add them here and add extra sparse checks

zou3519 replied:
I added four of the sparse dispatch keys. There are three that I don't know if I should have added:

https://github.com/pytorch/pytorch/blob/baac51ff4ad492c3c89d4b189a57543a89ae2b3d/c10/core/DispatchKey.h#L111-L114

SparseHIP seems to be dead (free dispatch key?) and SparseXPU/SparseVE are out of tree so I'd want their developers to come to us and say they want this feature before opting them into it.

At any rate we can always add more to the list

albanD commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -43,7 +34,7 @@ bool resize_output(const Tensor& output, IntArrayRef shape) {
     // TODO(#61485): functorch wrapped tensors should not go through the
     // fast path. This is a hack, longer term solutions are in the issue
-    if (output.is_cpu() && !is_functorch_wrapped_tensor(output)) {
+    if (output.is_cpu() && !isFuncTorchWrappedTensor(output)) {
```
Why can't this be a subclass like check?

zou3519 replied:
It can

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -205,7 +207,13 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+      // optimization: use in-place
+      is_eq_max = (self.abs() == norm).logical_or_(self.isnan().logical_and_(norm.isnan())).type_as(self);
+    } else {
+      is_eq_max = (self.abs() == norm).logical_or_(self.isnan().logical_and(norm.isnan())).type_as(self);
```
the or is still inplace though?

zou3519 replied:
Oops, thanks for the catch

zou3519 replied:
Actually, I'm not sure if this is a problem. It's actually OK that the logical_or_ is in-place, because `self.isnan().logical_and(norm.isnan())` produces a Tensor with both the characteristics of `self` and `norm`

albanD replied:
You mean `self.abs() == norm` ?

zou3519 replied:
`self.abs() == norm` gives a tensor with both the characteristics of self and norm; if one is batched then the final result is batched.

`self.isnan().logical_and(norm.isnan())` does the same thing

So in the end, 
```
(self.abs() == norm) .inplace_( self.isnan().logical_and(norm.isnan()) )
```
is okay

albanD replied:
I agree that this is fine.

Just want to confirm that the only one that is important here is `self.abs() == nor` (the one being modified inplace) but we don't care aobut `self.isnan().logical_and(norm.isnan())` (even though in this particular case it is always fine)?

zou3519 replied:
They're both important I think -- we're doing `a.inplace_(b)` where `a` is `self.abs() == norm` and `b` is `self.isnan().logical_and(norm.isnan())`. In order for this to work, if `b` is e.g. a BatchedTensor, then `a` must be one too

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -947,13 +969,13 @@ Tensor native_dropout_double_backward(const Tensor& ggI, const Tensor& grad, con
     auto mask = value.isnan().item<bool>() ? input.isnan() : input == value;
     return grad.new_zeros(input.sizes(), input.options()).masked_fill_(mask, grad / mask.sum());
   }
+  auto mask = (input == value).logical_or_(input.isnan().logical_and_(value.isnan()));
```
this is also doing inplace. That's fine for you?

zou3519 replied:
Yes. Let's assume that input is a subclass A. Then `value` is also a subclass A (because value is the result of the operation in the forward pass). We are able to do (A == A).inplace_(A.inplace_(A)).

zou3519 replied:
On second thought, this works for vmap, but I don't know if `A.inplace_(A)` works for all Tensor subclasses

zou3519 replied:
Proposal: a subclass A that does not support (a: A).inplace_(b: A) in cases where regular Tensors do should rely on always-on functionalization to solve their problem

albanD replied:
Note that the `input.isnan().logical_and_(...` does still sounds problematic if value is batched but not input?

zou3519 replied:
Updated

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -656,7 +664,11 @@ Tensor clamp_backward(const Tensor & grad, const Tensor &self, const Tensor& min
-    return where((self >= min).logical_and_(self <= max), grad, zero);
+    if (!areAnyTensorSubclassLike({self, min, max})) {
+      // fastpath
+      return where((self >= min).logical_and_(self <= max), grad, zero);
```
We could reduce code duplication by having intermediary result and ` ? : ;` kind of checks no?

zou3519 replied:
Yup

zou3519 committed:
Update on "Introduce isTensorSubclassLike; add special cases to backwards formulas"


Something is TensorSubclassLike if it is a Tensor subclass or if it has
the same problems as Tensor subclasses. Today that just includes Tensor
Subclasses and meta tensors but may include other things in the future.

Some of our backwards formulas are incompatible with TensorSubclassLike
objects. For example, calling .data_ptr() is a problem because many
TensorSubclassLike objects don't have storage. Another problem is
in-place operations: performing `regular_tensor.inplace_(tensor_subclass)`
is a problem.

This PR adds special cases to the backward formulas for torch.max and
torch.clamp to handle this. The backward formulas for torch.max and
torch.clamp are not dispatcher operations so they cannot be overridden
and we hesitate to make them dispatcher operations for FC/BC concerns
and performance overhead concerns.

Furthermore, the old concept of "is this inplace operation vmap
compatible?" can be subsumed by the general "is this inplace operation
tensor-subclass compatible" question, so I replaced all instances of
isInplaceVmapCompatible and replaced it with the isTensorSubclassLike
checks.

Test Plan
- I tested the changes using functorch.
- It's possible to write a test for these in core (one has to make
a custom tensor subclass and then send it through the operation and then
invoke autograd), but I wanted to push the work to doing some
generic testing for backward formulas
(https://github.com/pytorch/pytorch/issues/69530) instead of doing some
one-off things now.

[ghstack-poisoned]

zou3519 commented:
Updated:
- added sparse keys to the check
- Tried to do some code deduplication. I'm not sure I like the result because it is more LOC, but the code was indeed deduplicated

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -205,7 +207,16 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+    Tensor self_and_norm_isnan;
+    const auto self_isnan = self.isnan();
+    const auto norm_isnan = norm.isnan();
+    if (!areAnyTensorSubclassLike({self, norm})) {
```
could this be 
```cpp
auto self_and_norm_isnan = areAnyTensorSubclassLike({self, norm}) ?
    self_isnan.logical_and(norm_isnan) : 
    self_isnan.logical_and_(norm_isnan);
```
?

zou3519 replied:
That doesn't avoid the refcount: `auto` resolves to `Tensor` and never a reference type

zou3519 replied:
The way to avoid the refcount bump is to have a lot of code duplication like there was before, let me know if you would prefer that.

albanD replied:
I does avoid one refcount compared to the code above where you first create a (empty) Tensor, then assign to it.
We could also do `const auto&` here and rely on lifetime extension to keep the returned Tensor alive in case it is fresh?

zou3519 replied:
Today I learned about lifetime extension: https://abseil.io/tips/107

zou3519 replied:
> I does avoid one refcount compared to the code above where you first create a (empty) Tensor, then assign to it.

Oh I see, that's a great point. Thanks for the clarification

zou3519 committed:
Update on "Introduce isTensorSubclassLike; add special cases to backwards formulas"


Something is TensorSubclassLike if it is a Tensor subclass or if it has
the same problems as Tensor subclasses. Today that just includes Tensor
Subclasses and meta tensors but may include other things in the future.

Some of our backwards formulas are incompatible with TensorSubclassLike
objects. For example, calling .data_ptr() is a problem because many
TensorSubclassLike objects don't have storage. Another problem is
in-place operations: performing `regular_tensor.inplace_(tensor_subclass)`
is a problem.

This PR adds special cases to the backward formulas for torch.max and
torch.clamp to handle this. The backward formulas for torch.max and
torch.clamp are not dispatcher operations so they cannot be overridden
and we hesitate to make them dispatcher operations for FC/BC concerns
and performance overhead concerns.

Furthermore, the old concept of "is this inplace operation vmap
compatible?" can be subsumed by the general "is this inplace operation
tensor-subclass compatible" question, so I replaced all instances of
isInplaceVmapCompatible and replaced it with the isTensorSubclassLike
checks.

Test Plan
- I tested the changes using functorch.
- It's possible to write a test for these in core (one has to make
a custom tensor subclass and then send it through the operation and then
invoke autograd), but I wanted to push the work to doing some
generic testing for backward formulas
(https://github.com/pytorch/pytorch/issues/69530) instead of doing some
one-off things now.

[ghstack-poisoned]

albanD reviewed (approved):
Update looks good

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32967727).
